### PR TITLE
add safeivdep

### PR DESCRIPTION
**Pull request by chriselrod**

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by Tokazama** [comment]
Is this intended to be specific to compiler directives, or will it be used to more generally inform parallel operations? If it's intended for more general use I'd suggest suggest something more clearly meaningful. I'm probably not a good judge of quality on this because I had to look this up but I imagine "ignore_vector_dependencies" would be too vague. Maybe `can_parallel_mutate`?

**Comment by chriselrod** [comment]
`indices_do_not_alias`?
And for the element type `instances_do_not_alias`?

**Comment by Tokazama** [comment]
Sounds good to me

**Review by Tokazama — APPROVED** [review]

File: lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl [hunk 1]
```diff
@@ -539,4 +539,30 @@ ndims_index(@nospecialize T::Type{<:Base.LogicalIndex}) = ndims(fieldtype(T, :ma
...
+That is, would it be safe to write to an array of type `T` in parallel?
+Examples where this is not true are `BitArray`s or `view(rand(6), [1,2,3,1,2,3])`.
+That is, it is not safe whenever different indices may alias the same memory.
+"""
+indices_do_not_alias(::Type) = false
+indices_do_not_alias(::Type{A}) where {T, A<:Base.StridedArray{T}} = instances_do_not_alias(T)
+indices_do_not_alias(::Type{Adjoint{T,A}}) where {T, A <: AbstractArray{T}} = indices_do_not_alias(A)
+indices_do_not_alias(::Type{Transpose{T,A}}) where {T, A <: AbstractArray{T}} = indices_do_not_alias(A)
+indices_do_not_alias(::Type{<:SubArray{<:Any,<:Any,A,I}}) where {
+  A,I<:Tuple{Vararg{Union{Base.RangeIndex, Base.ReshapedUnitRange, Base.AbstractCartesianIndex}}}} = indices_do_not_alias(A)
```
Comment (N5N3) [inline lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:566]: This seems not safe is some edge case?
```julia
0 .* (1:10) isa Base.RangeIndex # true
```

Comment (chriselrod) [reply lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:566]: True. This should probably be `UnitRange`.

Comment (chriselrod) [reply lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:566]: https://github.com/JuliaArrays/ArrayInterface.jl/pull/294

**Pull request merged**